[pt2e][quant] Update some docs for pt2 export quantization (#2530)
pytorch/tutorials: prototype_source/pt2e_quant_ptq_static.rst

.. code:: python

> 

Reviewer: I can't comment up there by in L25 it still says "Dynamo Export." Should we change this to just "Export" since dynamo is hidden now?

Reviewer: Also I noticed in L193 we're still importing `prepare_fx`, `convert_fx` etc. I guess these were incorrectly copied over from the other tutorials

.. code-block:: python

    from torch._export import capture_pre_autograd_graph

    example_inputs = (torch.rand(2, 3, 224, 224),)

>     exported_model, _ = capture_pre_autograd_graph(model_to_quantize, *example_inputs)

Reviewer: Do we need to add the dynamic input constraints here to represent "symbolic"?

Reviewer: Also, there's an extra `_` here; `capture_pre_autograd_graph` returns a single item, not a tuple

Author: we don't need "symbolic", but people can use constraints to express this, I can add a comment for dynamic_dim

    print(quantized_model)

.. note::
   At this step, we currently have two representations that you can choose from, but what exact representation
   we offer in the long term might change based on feedbacks from users.

   * Q/DQ Representation (default)
   Previous documentation for `representations <https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md>`_ all quantized operators are represented as ``dequantize -> fp32_op -> qauntize``.

   .. code-block:: python

>       def quantized_linear(x_int8, x_scale, x_zero_point, weight_int8, weight_scale, weight_zero_point, bias_fp32, output_scale, output_zero_point):

Reviewer 2: ```suggestion

      def quantized_linear(x_int8, x_scale, x_zero_point, weight_int8, weight_scale, weight_zero_point, bias_fp32, output_scale, output_zero_point):
```

Reviewer 2: Otherwise, the code block won't render

          x_fp32 = torch.ops.quantized_decomposed.dequantize_per_tensor(
                   x_i8, x_scale, x_zero_point, x_quant_min, x_quant_max, torch.int8)
          weight_fp32 = torch.ops.quantized_decomposed.dequantize_per_tensor(
                   weight_i8, weight_scale, weight_zero_point, weight_quant_min, weight_quant_max, torch.int8)
          weight_permuted = torch.ops.aten.permute_copy.default(weight_fp32, [1, 0]);
          out_fp32 = torch.ops.aten.addmm.default(bias_fp32, x_fp32, weight_permuted)
          out_i8 = torch.ops.quantized_decomposed.quantize_per_tensor(
          out_fp32, out_scale, out_zero_point, out_quant_min, out_quant_max, torch.int8)
          return out_i8
     

>      * Reference Quantized Model Representation (WIP, expected to be ready at end of August): we have special representation for selected ops (e.g. quantized linear), other ops are represented as (dq -> float32_op -> q), and q/dq are decomposed into more primitive operators.

Reviewer 2: ```suggestion
     * Reference Quantized Model Representation (WIP, expected to be ready at end of August): we have special representation for selected ops (for example, quantized linear), other ops are represented as (dq -> float32_op -> q), and q/dq are decomposed into more primitive operators.
```

>        You can get this representation by: convert_pt2e(..., use_reference_representation=True)

Reviewer 2: ```suggestion
       You can get this representation by: ``convert_pt2e(..., use_reference_representation=True)``
```

    .. code-block:: python
       # Reference Quantized Pattern for quantized linear
       def quantized_linear(x_int8, x_scale, x_zero_point, weight_int8, weight_scale, weight_zero_point, bias_fp32, output_scale, output_zero_point):
           x_int16 = x_int8.to(torch.int16)
           weight_int16 = weight_int8.to(torch.int16)
           acc_int32 = torch.ops.out_dtype(torch.mm, torch.int32, (x_int16 - x_zero_point), (weight_int16 - weight_zero_point))
           acc_rescaled_int32 = torch.ops.out_dtype(torch.ops.aten.mul.Scalar, torch.int32, acc_int32, x_scale * weight_scale / output_scale)
           bias_scale = x_scale * weight_scale
           bias_int32 = out_dtype(torch.ops.aten.mul.Tensor, torch.int32, bias_fp32, bias_scale / out_scale)
           out_int8 = torch.ops.aten.clamp(acc_rescaled_int32 + bias_int32 + output_zero_point, qmin, qmax).to(torch.int8)
           return out_int8

>    Please see `<here https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/pt2e/representation/rewrite.py>`_ for the most up to date reference representations.

Reviewer 2: ```suggestion
   See `here <https://github.com/pytorch/pytorch/blob/main/torch/ao/quantization/pt2e/representation/rewrite.py>`_ for the most up-to-date reference representations.
```

Reviewer 2: This is right now not rendering correctly

    print(quantized_model)

.. note::

>    At this step, we currently have two representations that you can choose from, but what exact representation
>    we offer in the long term might change based on feedbacks from users.

Reviewer 2: ```suggestion
   At this step, we currently have two representations that you can choose from, but exact representation
   we offer in the long term might change based on feedback from PyTorch users.
```

>    * Q/DQ Representation (default)
>    Previous documentation for `representations <https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md>`_ all quantized operators are represented as ``dequantize -> fp32_op -> qauntize``.

Reviewer 2: ```suggestion
   * Q/DQ Representation (default)
      
      Previous documentation for `representations <https://github.com/pytorch/rfcs/blob/master/RFC-0019- 
 Extending-PyTorch-Quantization-to-Custom-Backends.md>`_ all quantized operators are represented as ``dequantize -> fp32_op -> qauntize``.
```

   .. code-block:: python

      def quantized_linear(x_int8, x_scale, x_zero_point, weight_int8, weight_scale, weight_zero_point, bias_fp32, output_scale, output_zero_point):
          x_fp32 = torch.ops.quantized_decomposed.dequantize_per_tensor(
                   x_i8, x_scale, x_zero_point, x_quant_min, x_quant_max, torch.int8)
          weight_fp32 = torch.ops.quantized_decomposed.dequantize_per_tensor(
                   weight_i8, weight_scale, weight_zero_point, weight_quant_min, weight_quant_max, torch.int8)
          weight_permuted = torch.ops.aten.permute_copy.default(weight_fp32, [1, 0]);
          out_fp32 = torch.ops.aten.addmm.default(bias_fp32, x_fp32, weight_permuted)
          out_i8 = torch.ops.quantized_decomposed.quantize_per_tensor(
          out_fp32, out_scale, out_zero_point, out_quant_min, out_quant_max, torch.int8)
          return out_i8
     

>      * Reference Quantized Model Representation (WIP, expected to be ready at end of August): we have special representation for selected ops (for example, quantized linear), other ops are represented as (dq -> float32_op -> q), and q/dq are decomposed into more primitive operators.

Reviewer 2: ```suggestion
     * Reference Quantized Model Representation (WIP, expected to be ready at end of August): we have special representation for selected ops (for example, quantized linear), other ops are represented as (``dq -> float32_op -> q``), and ``q/dq`` are decomposed into more primitive operators.
```

Reviewer 2: The output is broken for this at the moment: https://docs-preview.pytorch.org/pytorch/tutorials/2530/prototype/pt2e_quant_ptq_static.html#convert-the-calibrated-model-to-a-quantized-model

Author: I see, I'll fix it

>        You can get this representation by: ``convert_pt2e(..., use_reference_representation=True)``

Reviewer 2: ```suggestion
       You can get this representation by using ``convert_pt2e(..., use_reference_representation=True)``.
```

>     .. code-block:: python

Reviewer 2: ```suggestion
    .. code-block:: python
    
```
